core-plugins/src/main/java/io/cdap/plugin/batch/joiner/Joiner.java

```diff
@@ -239,37 +251,42 @@ void validateJoinKeySchemas(Map<String, Schema> inputSchemas, Map<String, List<S
 
       Schema schema = inputSchemas.get(stageName);
       if (schema == null) {
-        throw new IllegalArgumentException(String.format("Input schema for input stage %s can not be null", stageName));
+        collector.addFailure(String.format("Input schema for input stage %s cannot be null", stageName), null);
+        collector.getOrThrowException();
       }
 
       for (String joinKey : entry.getValue()) {
         Schema.Field field = schema.getField(joinKey);
         if (field == null) {
-          throw new IllegalArgumentException(String.format("Join key field %s is not present in input of stage %s",
-                                                           joinKey, stageName));
+          collector.addFailure(
+              String.format("Join key field %s is not present in input stage of %s", joinKey, stageName), null)
+              .withConfigProperty(JOIN_KEYS);
+          collector.getOrThrowException();
         }
         schemaList.add(field.getSchema());
       }
       if (prevSchemaList != null && !prevSchemaList.equals(schemaList)) {
-        throw new IllegalArgumentException(String.format("For stage %s, Schemas of joinKeys %s are expected to be: " +
-                                                           "%s, but found: %s",
-                                                         stageName, entry.getValue(), prevSchemaList.toString(),
-                                                         schemaList.toString()));
+        collector.addFailure(
+            String.format("For stage %s, Schemas of joinKeys %s are expected to be: %s, but found: %s",
```

Comment:
you'll see this in some older plugins, but error messages shouldn't reference internal variable names or classes like this. Instead, they should use plain english:

Schemas of joinKeys -> schemas of join keys

Reply:
Fixed